docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+Custom Attestation Data Formats for Open Enclave V3 Updates
+================
+
+As custom attestation plugins are developed and features are requested, new APIs were added to:
+
+1. Query built-in SGX attesters and verifiers to be registered.
+2. Select a format of the evidence that the registered attesters can generate.
+3. Query evidence parameters to be used as optional inputs when generating evidences.
+
+The following API was affected:
+
+1. Built-in SGX attesters and verifiers
+
+This document describes the added and affected APIs as well as keeping the existing support for version 2 of the API. This document assumes the user is familiar with the SGX attestation and [custom attestation](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation.md) with plugins.
+
+Motivation
+----------
+
+New APIs were added to help customers who would like to extend Open Enclave's current attestation structure with extra information while keep TEE-specific information within attester and verifier plugins.
+
+1. The application can query built-in SGX attesters and verifiers to be registered. This allows the plugin library to return attesters and verifiers based on configuration.
```

Comment:
This is a replacement of existing SGX-specific API. In current release, people can register SGX attesters and verifiers. The same functionality remains but we are making it more dynamic.

---
In reply to: [393246199](https://github.com/openenclave/openenclave/pull/2621#discussion_r393246199) [](ancestors = 393246199)

Reply:
If TEE-specific API is required to avoid need to force-link, one option to make the API TEE-agnostic to application software is for OE SDK header to have a macro to rename a TEE-specific function to a TEE-agnostic function name. OE SDK header would need to provide different macros for different platforms, but application software would stay the same across those platforms.

Reply:
This works fine for application with one plugin. But with multiple plugins, the macro approach would result in a naming conflict.

---
In reply to: [393880911](https://github.com/openenclave/openenclave/pull/2621#discussion_r393880911) [](ancestors = 393880911)